File: check_circular_cll.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

Node* push(Node* head, int data) {
    Node* newn = (Node*)malloc(sizeof(Node));
    newn->data = data;
    newn->next = head;
    return newn;
}

// Detect cycle using Floyd
int hasCycle(Node* head) {
    Node *slow = head, *fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return 1;
    }
    return 0;
}

// Check structural circular where traversal returns to head
int isCircularList(Node* head) {
    if (!head) return 1; // empty considered circular here
    Node* cur = head->next;
    while (cur && cur != head) cur = cur->next;
    return (cur == head);
}

int main() {
    Node* head = NULL;
    head = push(head, 3);
    head = push(head, 2);
    head = push(head, 1); // simple list 1->2->3->NULL

    printf("Has cycle? %s\n", hasCycle(head) ? "Yes" : "No");
    printf("Is structurally circular? %s\n", isCircularList(head) ? "Yes" : "No");

    // Make it circular: 3->head
    Node* tail = head;
    while (tail->next) tail = tail->next;
    tail->next = head;

    printf("After making circular:\n");
    printf("Has cycle? %s\n", hasCycle(head) ? "Yes" : "No");
    printf("Is structurally circular? %s\n", isCircularList(head) ? "Yes" : "No");

    return 0;
}

File: check_circular_cll.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def has_cycle(head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

def is_circular_list(head):
    if not head:
        return True
    cur = head.next
    while cur and cur != head:
        cur = cur.next
    return cur == head

if __name__ == "__main__":
    head = Node(1)
    head.next = Node(2)
    head.next.next = Node(3)
    print("Has cycle?", has_cycle(head))
    print("Is structurally circular?", is_circular_list(head))

    # make circular
    tail = head
    while tail.next: tail = tail.next
    tail.next = head
    print("After making circular:")
    print("Has cycle?", has_cycle(head))
    print("Is structurally circular?", is_circular_list(head))
